src/emailjs-tcp-socket.js

```diff
@@ -619,8 +662,16 @@
                 // process the data available on the socket
                 self._onData(readInfo.data);
 
-                // queue the next read
-                self._readLegacySocket();
+                // Queue the next read.
+                // If a STARTTLS handshake might be upcoming, postpone this onto
+                // the task queue so the IMAP client has a chance to call upgradeToSecure;
+                // without this, we might eat the beginning of the handshake.
+                // If we are already secure, just call it (for performance).
+                if (self.ssl) {     // are we secure yet?
+                    self._readLegacySocket();
+                } else {
+                    setZeroTimeout(self._readLegacySocket.bind(self));
```

Comment:
Re TCP sockets on the open web. I had a long discussion with ryan sleevi and others from the chrome/mozilla network/ssl stack on twitter. It doesn't sound like we'll get sockets anytime soon. Mainly because of security of running untrusted code. Even with the permissions api, there is currently no good ux solution for asking average users for tcp access... "Allow this app to talk to the toaster in your kitchen?"